### PR TITLE
EZ-301: Cleanup - Edit Article - Edit Media - Edit image - hide additional fields

### PR DESCRIPTION
**Pull request by saurabh-2k17**
EZ-301: Cleanup - Edit Article - Edit Media - Edit image - hide additional fields

### CODE REVIEW

File: ezcontent.links.menu.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+ezcontent.settings_form:
```
Comment (vaibhavjain-in) [inline ezcontent.links.menu.yml:1]: Why is this file required ?

File: ezcontent.profile [hunk 2]
```diff
@@ -91,3 +94,17 @@ function ezcontent_themes_installed($theme_list) {
     }
   }
 }
+
+/**
+ * Implements hook_form_alter().
+ */
+function ezcontent_form_alter(array &$form, FormStateInterface $form_state, $form_id) {
+  if ($form_id == 'media_image_add_form' || $form_id == 'media_image_edit_form' || $form_id == 'media_instagram_add_form' || $form_id == 'media_instagram_edit_form' || $form_id == 'media_tweet_add_form' || $form_id == 'media_tweet_edit_form' || $form_id == 'media_video_add_form' || $form_id == 'media_video_edit_form') {
```
Comment (vaibhavjain-in) [inline ezcontent.profile]: This will fail in case there are more media bundles are added. Please see if this can be generalised.

File: ezcontent.profile [hunk 3]
```diff
@@ -91,3 +94,17 @@ function ezcontent_themes_installed($theme_list) {
...
+    $media_type = \Drupal::routeMatch()->getParameter('media_type');
```
Comment (vaibhavjain-in) [inline ezcontent.profile]: Can you get this from the form object ?

File: ezcontent.profile [hunk 4]
```diff
@@ -91,3 +94,17 @@ function ezcontent_themes_installed($theme_list) {
...
+    $media_name = $media_type->id();
+    $hide_status = \Drupal::config('ezcontent.settings')->get($media_name);
+    if ($hide_status === 1) {
```
Comment (vaibhavjain-in) [inline ezcontent.profile]:
```suggestion
    if ((bool) $hide_status === TRUE) {
```
OR
```suggestion
    if ($hide_status = \Drupal::config('ezcontent.settings')->get($media_name);) {
```

Also, inject the config object.

File: ezcontent.routing.yml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+ezcontent.settings_form:
+  path: '/admin/config/ezcontent/settings'
```
Comment (vaibhavjain-in) [inline ezcontent.routing.yml]:
```suggestion
  path: '/admin/config/system/ezcontent/settings'
```

File: src/Form/SettingsForm.php [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    $config = $this->config('ezcontent.settings');
+
+    $form['revision'] = [
+      '#type'  => 'fieldset',
+      '#title' => 'Hide Entity Revision Field',
+      '#open'  => TRUE,
+    ];
+
+    $form['revision']['media'] = [
+      '#type'  => 'fieldset',
```
Comment (vaibhavjain-in) [inline src/Form/SettingsForm.php]: Please make it collapsible and default as `collapsed`.
Check FVM settings at path `admin/structure/display-modes/view/fvm`.

File: src/Form/SettingsForm.php [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      '#title' => 'Media',
+      '#description' => $this->t('Please select media type for which revision information will be disabled.'),
```
Comment (vaibhavjain-in) [inline src/Form/SettingsForm.php]:
```suggestion
      '#description' => $this->t('Please select media bundle.'),
```

File: src/Form/SettingsForm.php [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      '#open'  => TRUE,
+    ];
+
+    $media_types = \Drupal::entityTypeManager()->getStorage('media_type')->loadMultiple();
```
Comment (vaibhavjain-in) [inline src/Form/SettingsForm.php]:
```suggestion
    $mediaTypes = \Drupal::entityTypeManager()->getStorage('media_type')->loadMultiple();
```

File: src/Form/SettingsForm.php [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    foreach ($media_types as $media_type_name => $value) {
```
Comment (vaibhavjain-in) [inline src/Form/SettingsForm.php]:
```suggestion
    foreach ($mediaTypes as $bundle => $entity) {
```

File: src/Form/SettingsForm.php [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      $form['revision']['media'][$media_type_name] = [
```
Comment (vaibhavjain-in) [inline src/Form/SettingsForm.php]:
```suggestion
      $form['revision']['media'][$bundle] = [
```

File: src/Form/SettingsForm.php [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        '#type' => 'checkbox',
+        '#title' => $this->t($media_type_name),
```
Comment (vaibhavjain-in) [inline src/Form/SettingsForm.php]:
```suggestion
        '#title' => $bundle->label(),
```

File: src/Form/SettingsForm.php [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+        '#default_value' => $config->get($media_type_name),
```
Comment (vaibhavjain-in) [inline src/Form/SettingsForm.php]:
```suggestion
        '#default_value' => $config->get($bundle),
```

File: src/Form/SettingsForm.php [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+      ];
+
```
Comment (vaibhavjain-in) [inline src/Form/SettingsForm.php]: Remove this extra line.

File: src/Form/SettingsForm.php [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    parent::validateForm($form, $form_state);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
```
Comment (vaibhavjain-in) [inline src/Form/SettingsForm.php:110]: Do we need this method to be overwritten ?

**Review by vaibhavjain-in — CHANGES_REQUESTED** [review]
@saurabh-2k17 Please cater to feedback.

File: ezcontent.links.menu.yml (hunk 1 above)
Comment (saurabh-2k17) [reply ezcontent.links.menu.yml:1]: For menu. I have described its parent.

File: ezcontent.profile (hunk 3 above)
Comment (saurabh-2k17) [reply ezcontent.profile]: Sir, it is not available in $form obj.

File: ezcontent.links.menu.yml [hunk 15]
```diff
@@ -0,0 +1,6 @@
...
+  title: 'EZContent Config Form'
```
Comment (vaibhavjain-in) [inline ezcontent.links.menu.yml]:
```suggestion
  title: 'EZContent Settings'
```

File: ezcontent.links.menu.yml [hunk 16]
```diff
@@ -0,0 +1,6 @@
...
+  route_name: ezcontent.settings_form
+  description: 'Menu for EZContent config form'
```
Comment (vaibhavjain-in) [inline ezcontent.links.menu.yml]:
```suggestion
  description: 'Menu for EZContent settings form'
```

File: ezcontent.links.menu.yml [hunk 17]
```diff
@@ -0,0 +1,6 @@
...
+  parent: system.admin_config_system
+  weight: 99
```
Comment (vaibhavjain-in) [inline ezcontent.links.menu.yml]: Please see if we can remove the weight key.

File: ezcontent.profile [hunk 18]
```diff
@@ -91,3 +94,20 @@ function ezcontent_themes_installed($theme_list) {
     }
   }
 }
+
+/**
+ * Implements hook_form_alter().
+ */
+function ezcontent_form_alter(array &$form, FormStateInterface $form_state, $form_id) {
+  $route = \Drupal::routeMatch()->getParameters()->get('entity_type_id');
+  if ($route == 'media') {
```
Comment (vaibhavjain-in) [inline ezcontent.profile]: Lets use switch statement here.

Comment (vaibhavjain-in) [reply ezcontent.profile]: Also, rather than validating from $route object, check via `$form_id`

File: ezcontent.routing.yml [hunk 19]
```diff
@@ -0,0 +1,9 @@
+ezcontent.settings_form:
+  path: '/admin/config/system/ezcontent/settings'
+  defaults:
+    _form: '\Drupal\ezcontent\Form\SettingsForm'
+    _title: 'EZContent Settings Form'
```
Comment (vaibhavjain-in) [inline ezcontent.routing.yml]:
```suggestion
    _title: 'EZContent Settings'
```

File: src/Form/SettingsForm.php [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+    return [
+      'ezcontent.settings',
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFormId() {
+    return 'settings_form';
```
Comment (vaibhavjain-in) [inline src/Form/SettingsForm.php]: lets create a unique name `ezcontent_settings_form`

File: src/Form/SettingsForm.php [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $config = $this->config('ezcontent.settings');
+
+    $form['revision'] = [
+      '#type'  => 'fieldset',
+      '#title' => 'Hide Entity Revision Field',
```
Comment (vaibhavjain-in) [inline src/Form/SettingsForm.php]:
```suggestion
      '#title' => 'Cleanup fields in Entity add / edit form',
```
Also, add desc

File: src/Form/SettingsForm.php (hunk 6 above)
Comment (vaibhavjain-in) [reply src/Form/SettingsForm.php]: Please use type `details`

File: src/Form/SettingsForm.php [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+      '#title' => 'Media',
+      '#description' => $this->t('Please select media bundle.'),
+      '#open'  => TRUE,
+    ];
+
+    $mediaTypes = \Drupal::entityTypeManager()->getStorage('media_type')->loadMultiple();
+    foreach ($mediaTypes as $bundle => $entity) {
+      $form['revision']['media'][$bundle] = [
+        '#type' => 'checkbox',
+        '#title' => $bundle,
```
Comment (vaibhavjain-in) [inline src/Form/SettingsForm.php]:
```suggestion
        '#title' => $entity->label(),
```

**Review by vaibhavjain-in — CHANGES_REQUESTED** [review]

File: ezcontent.profile [hunk 23]
```diff
@@ -91,3 +94,20 @@ function ezcontent_themes_installed($theme_list) {
     }
   }
 }
+
+/**
+ * Implements hook_form_alter().
+ */
+function ezcontent_form_alter(array &$form, FormStateInterface $form_state, $form_id) {
+  $route = \Drupal::routeMatch()->getParameters()->get('entity_type_id');
```
Comment (arshadkhan53) [inline ezcontent.profile]: name should be $entityTypeId instead of $route

File: src/Form/SettingsForm.php [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildForm(array $form, FormStateInterface $form_state) {
+    $config = $this->config('ezcontent.settings');
+
+    $form['revision'] = [
+      '#type'  => 'fieldset',
```
Comment (arshadkhan53) [inline src/Form/SettingsForm.php]: I believe you can use details field type here

File: src/Form/SettingsForm.php [hunk 25]
```diff
@@ -0,0 +1,83 @@
...
+    parent::validateForm($form, $form_state);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitForm(array &$form, FormStateInterface $form_state) {
+    parent::submitForm($form, $form_state);
+
+    $media_types = \Drupal::entityTypeManager()->getStorage('media_type')->loadMultiple();
```
Comment (arshadkhan53) [inline src/Form/SettingsForm.php]: Please DI in here

File: src/Form/SettingsForm.php [hunk 26]
```diff
@@ -0,0 +1,83 @@
...
+    ];
+
+    $form['revision']['media'] = [
+      '#type'  => 'fieldset',
+      '#title' => 'Media',
+      '#description' => $this->t('Please select media bundle.'),
+      '#open'  => TRUE,
+    ];
+
+    $mediaTypes = \Drupal::entityTypeManager()->getStorage('media_type')->loadMultiple();
```
Comment (arshadkhan53) [inline src/Form/SettingsForm.php]: Please use DI here

File: src/Form/SettingsForm.php (hunk 14 above)
Comment (saurabh-2k17) [reply src/Form/SettingsForm.php:110]: Yes sir in order to keep the values on form

File: ezcontent.profile [hunk 27]
```diff
@@ -91,3 +94,23 @@ function ezcontent_themes_installed($theme_list) {
...
+      case "media":
+        if (strpos($form_id, 'media_'.$mediaType) !== FALSE) {
+          $hide_status = \Drupal::config('ezcontent.settings')->get($mediaType);
+          if ((bool) $hide_status === TRUE) {
+            $form['revision_information']['#access'] = FALSE;
+          }    
+        }
+      break;
+    }
+  }
```
Comment (vaibhavjain-in) [inline ezcontent.profile]:
```suggestion
  if ((strpos($form_id, '_add_form') || (strpos($form_id, '_edit_form'))) !== FALSE) {
    /* @var $entity Drupal\Core\Entity\EditorialContentEntityBase */
    $entity = $form_state->getFormObject()->getEntity();
    if (is_subclass_of($entity, 'Drupal\Core\Entity\EditorialContentEntityBase')) {
      switch ($entity->getEntityTypeId()) {
        case "media":
          if (strpos($form_id, 'media_' . $entity->bundle()) !== FALSE) {
            if ((bool) \Drupal::config('ezcontent.settings')
                ->get($entity->bundle()) === TRUE) {
              $form['revision_information']['#access'] = FALSE;
            }
          }
          break;
      }
    }
  }
```

File: src/Form/SettingsForm.php [hunk 28]
```diff
@@ -0,0 +1,104 @@
+<?php
```
Comment (vaibhavjain-in) [inline src/Form/SettingsForm.php:1]: @arshadkhan53 Good number of changes here, hence pasting the code of full file below.

```<?php

namespace Drupal\ezcontent\Form;

use Drupal\Core\Entity\EntityTypeManagerInterface;
use Drupal\Core\Form\ConfigFormBase;
use Drupal\Core\Form\FormStateInterface;
use Symfony\Component\DependencyInjection\ContainerInterface;
use Drupal\Core\Config\ConfigFactoryInterface;
use Drupal\Core\Entity\EntityTypeBundleInfoInterface;

/**
 * Class SettingsForm.
 *
 * @package Drupal\ezcontent\Form
 */
class SettingsForm extends ConfigFormBase {

  /**
   * The entity type manager.
   *
   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
   */
  protected $entityTypeManager;

  /**
   * Bundle info.
   *
   * @var \Drupal\Core\Entity\EntityTypeBundleInfoInterface
   */
  protected $bundleInfo;

  /**
   * Constructs a \Drupal\ezcontent\SettingsForm object.
   *
   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
   *   The factory for configuration objects.
   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
   *   The entity type manager.
   * @param \Drupal\Core\Entity\EntityTypeBundleInfoInterface $bundleInfo
   *   Bundle info.
   */
  public function __construct(ConfigFactoryInterface $config_factory, EntityTypeManagerInterface $entityTypeManager, EntityTypeBundleInfoInterface $bundleInfo) {
    parent::__construct($config_factory);
    $this->entityTypeManager = $entityTypeManager;
    $this->bundleInfo = $bundleInfo;
  }

  /**
   * {@inheritdoc}
   */
  public static function create(ContainerInterface $container) {
    return new static(
      $container->get('config.factory'),
      $container->get('entity_type.manager'),
      $container->get('entity_type.bundle.info')
    );
  }

  /**
   * {@inheritdoc}
   */
  protected function getEditableConfigNames() {
    return [
      'ezcontent.settings',
    ];
  }

  /**
   * {@inheritdoc}
   */
  public function getFormId() {
    return 'ezcontent_settings_form';
  }

  /**
   * {@inheritdoc}
   */
  public function buildForm(array $form, FormStateInterface $form_state) {
    $config = $this->config('ezcontent.settings');
    $form['hide_field'] = [
      '#type'  => 'details',
      '#title' => $this->t('Cleanup fields in Entity add / edit form'),
      '#description' => $this->t('Hide fields, which cannot be hidden via manage form display section.'),
      '#open'  => TRUE,
    ];
    $entityTypes = $this->entityTypeManager->getDefinitions();
    foreach ($entityTypes as $entityId => $entityType) {
      if ($entityId == 'media') {
        $form['hide_field'][$entityId] = [
          '#type'  => 'details',
          '#title' => $entityType->getLabel(),
          '#description' => $this->t('Please select bundle.'),
          '#open'  => TRUE,
        ];

        foreach ($this->bundleInfo->getBundleInfo($entityId) as $bundleId => $bundle) {
          $form['hide_field'][$entityId][$bundleId] = [
            '#type' => 'checkbox',
            '#title' => $bundle['label'],
            '#default_value' => $config->get($bundleId),
          ];
        }
      }
    }

    return parent::buildForm($form, $form_state);
  }

  /**
   * {@inheritdoc}
   */
  public function submitForm(array &$form, FormStateInterface $form_state) {
    parent::submitForm($form, $form_state);

    $config = $this->config('ezcontent.settings');
    foreach ($form_state->cleanValues()->getValues() as $key => $value) {
      $config->set($key, $value);
    }
    $config->save();
  }

}
```

**Review by vaibhavjain-in — CHANGES_REQUESTED** [review]
@arshadkhan53 Please see if the changes are correct. I added all changes I made on my local and they seem to be working good.

File: ezcontent.links.menu.yml [hunk 29]
```diff
@@ -0,0 +1,5 @@
+ezcontent.settings_form:
+  title: 'EZContent Settings'
+  route_name: ezcontent.settings_form
```
Comment (arshadkhan53) [inline ezcontent.links.menu.yml:3]: Please rename EZContent to EzContent

File: ezcontent.links.menu.yml [hunk 30]
```diff
@@ -0,0 +1,5 @@
...
+  description: 'Menu for EZContent settings form'
```
Comment (arshadkhan53) [inline ezcontent.links.menu.yml]: Please rename EZContent to EzContent

**Review by arshadkhan53 — APPROVED** [review]
LGTM

**Review by vaibhavjain-in — APPROVED** [review]
LGTM
